tsinfer/formats.py

```diff
@@ -3088,21 +3102,57 @@ def truncate_ancestors(
                         f"Truncating ancestor {anc.id} at time {anc.time}"
                         "Original length {original_length}. New length {new_length}"
                     )
-                    start[anc.id] = insert_pos_start
-                    end[anc.id] = insert_pos_end
-                    time[anc.id] = anc.time
-                    focal_sites[anc.id] = anc.focal_sites
-                    haplotypes[:, anc.id] = tskit.MISSING_DATA
-                    haplotypes[
-                        insert_pos_start:insert_pos_end, anc.id, 0
+                    index_buffer[buffer_pos] = anc_index
+                    start_buffer[buffer_pos] = insert_pos_start
+                    end_buffer[buffer_pos] = insert_pos_end
+                    time_buffer[buffer_pos] = anc.time
+                    focal_sites_buffer[buffer_pos] = anc.focal_sites
+                    haplotype_buffer[
+                        insert_pos_start:insert_pos_end, buffer_pos, 0
                     ] = anc.full_haplotype[insert_pos_start:insert_pos_end]
-                    # TODO - record truncation in ancestors' metadata when supported
-        truncated.ancestors_start[:] = start
-        truncated.ancestors_end[:] = end
-        truncated.ancestors_time[:] = time
-        truncated.ancestors_focal_sites[:] = focal_sites
-        truncated.ancestors_full_haplotype[:] = haplotypes
-        truncated.ancestors_full_haplotype_mask[:] = haplotypes == tskit.MISSING_DATA
+                    buffer_pos += 1
+                    if buffer_pos == buffer_length:
```

Comment:
A local function would help here,

```python
def flush_buffer(length):
       truncated.ancestors_start.set_orthogonal_selection(
              index_buffer[:length], start_buffer[:length]
         )
        # etc
```

Reply:
I don't follow how this is working, so a few comments on how ``flush_buffer`` works and would be helpful here.